services/fileuploader/fileuploader.go

```diff
@@ -0,0 +1,160 @@
+package fileuploader
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/rudderlabs/rudder-server/config"
+	backendconfig "github.com/rudderlabs/rudder-server/config/backend-config"
+	"github.com/rudderlabs/rudder-server/services/filemanager"
+)
+
+type StorageSettings struct {
+	Bucket      backendconfig.StorageBucket
+	Preferences backendconfig.StoragePreferences
+}
+
+// Provider is an interface that provides file managers and storage preferences for a given workspace.
+type Provider interface {
+	// Gets a file manager for the given workspace.
+	GetFileManager(workspaceID string) (filemanager.FileManager, error)
+	// Gets the storage preferences for the given workspace.
+	GetStoragePreferences(workspaceID string) (backendconfig.StoragePreferences, error)
+}
+
+// NewProvider creates a new provider that updates its storage settings while backend configuration gets updated.
+func NewProvider(ctx context.Context, config backendconfig.BackendConfig) Provider {
+	s := &provider{
+		init:            make(chan struct{}),
+		storageSettings: make(map[string]StorageSettings),
+	}
+	go s.updateLoop(ctx, config)
+	return s
+}
+
+// NewStaticProvider creates a new provider that operates against a predefined storage settings.
+// Useful for tests.
+func NewStaticProvider(storageSettings map[string]StorageSettings) Provider {
+	s := &provider{
+		init:            make(chan struct{}),
+		storageSettings: storageSettings,
+	}
+	close(s.init)
+	return s
```

Comment:
please add a test scenario for this function

Reply:
Done